git clone using a keypair protected by a passphrase

- clj-jgit/clj-jgit {:mvn/version "1.1.0"}

I have a keypair generated with a passphrase and I have a keypair with no passphrase.

I can use the keypair without a passphase to clone a repo:
```clojure
(git/with-identity
  {:name      "no_passphrase"
   :trust-all true}
  (def my-repo (git/git-clone
                 "git@github.com:babashka/sci.git"
                 :dir "local-folder/sci")))
```

If I use the keypair with a passphase:
```clojure
(git/with-identity
  {:name      "id_rsa"
   :pw        "<secret>"
   :trust-all true}
  (def my-repo (git/git-clone
                 "git@github.com:babashka/sci.git"
                 :dir "local-folder/sci")))
```

I get an error:
`Caused by: javax.security.auth.login.FailedLoginException: No password data for encrypted resource=<path>/.ssh/id_rsa`

so I know the key file is found ok.

I see the `*ssh-key-passphrase*` binding correctly set in the debugger when the git-clone is executed but it doesn't seem to be used later.   Anybody have a working example or an idea please?

--- a/src/clj_jgit/porcelain.clj
+++ b/src/clj_jgit/porcelain.clj
@@ -26,7 +26,7 @@
            (org.eclipse.jgit.transport.sshd SshdSessionFactory DefaultProxyDataFactory JGitKeyCache KeyPasswordProvider)
            (org.eclipse.jgit.transport FetchResult UsernamePasswordCredentialsProvider URIish RefSpec RefLeaseSpec TagOpt
                                        RemoteConfig CredentialsProvider CredentialItem$CharArrayType
-                                       CredentialItem$YesNoType SshTransport)
+                                       CredentialItem$YesNoType CredentialItem$Password SshTransport)
            (org.eclipse.jgit.treewalk TreeWalk)
            (org.eclipse.jgit.treewalk.filter PathFilter)))
 
@@ -101,10 +101,13 @@
 #_{:clj-kondo/ignore [:unused-binding]}
 (defn key-pass-provider
   "Create a new `KeyPasswordProvider` instance for given `key-pw`."
-  ^KeyPasswordProvider [key-pw]
+  ^KeyPasswordProvider [^CredentialsProvider provider]
   (reify KeyPasswordProvider
     (getPassphrase [_ uri attempt]
-      (char-array key-pw))
+      (let [credential-item (CredentialItem$Password.)]
+        (if (.get provider uri [credential-item])
+          (.getValue credential-item)
+          nil)))
     (setAttempts [_ attempts]
       true)
     (getAttempts [_]
@@ -123,7 +126,7 @@
           ssh-dir
           (.getAbsoluteFile ssh-dir))))
     (createKeyPasswordProvider [^CredentialsProvider provider]
-      (key-pass-provider *ssh-key-passphrase*))
+      (key-pass-provider provider))
     (getDefaultKnownHostsFiles [^File ssh-dir]
       (let [ssh-dir (.toPath ssh-dir)
             kh-files (->> (seq?! *known-hosts-file*)
@@ -476,7 +479,9 @@
   (-> (Git/cloneRepository)
       (.setURI uri)
       ^TransportCommand (.setCredentialsProvider *cred-provider*)
-      (.setTransportConfigCallback *transport-callback*)))
+      (.setTransportConfigCallback *transport-callback*)
+      (.setCredentialsProvider(UsernamePasswordCredentialsProvider. nil *ssh-key-passphrase*))
+      ))
 
 (defn git-clone
   "Clone a repository into a new working directory from given `uri`.

--- a/src/clj_jgit/porcelain.clj
+++ b/src/clj_jgit/porcelain.clj
@@ -205,9 +205,10 @@
      (binding [*ssh-key-name* key-name#
                *ssh-key-dir* key-dir#
                *ssh-key-passphrase* key-pw#
-               *cred-provider* (if (and trust-all?# (nil? cred-provider#))
-                                 trust-any-provider
-                                 cred-provider#)
+               *cred-provider* (cond
+                                 (and trust-all?# (nil? cred-provider#)) trust-any-provider
+                                 (and key-pw# (nil? cred-provider#)) (user-pass-provider nil key-pw# :trust-all? trust-all?#)
+                                 :else cred-provider#)
                *transport-callback* transport-cb#]
        ~@body)))
 
@@ -479,9 +480,7 @@
   (-> (Git/cloneRepository)
       (.setURI uri)
       ^TransportCommand (.setCredentialsProvider *cred-provider*)
-      (.setTransportConfigCallback *transport-callback*)
-      (.setCredentialsProvider(UsernamePasswordCredentialsProvider. nil *ssh-key-passphrase*))
-      ))
+      (.setTransportConfigCallback *transport-callback*)))
 
 (defn git-clone
   "Clone a repository into a new working directory from given `uri`.

